### PR TITLE
Layer to convert tensor to ragged tensor dropping the ignored value

### LINKED ISSUE #1861: Support convert Tensor to RaggedTensor dropping the ignored value.
The format of the sequence is usually a string with separator such as the sentence in a text with whitespace separator. And the number of words in each sentence is not the same. We can split the sentences to a `tf.RaggedTensor`. Then we may want to drop some values like the empty string.

For example:
```python
data = np.array([["Hello world"], [""], ["Welcome to ElasticDL"]])
```
1. Split strings to 
```python
data = tf.RaggedTensor([["Hello", "world"], [""], ["Welcome", "to", "ElasticDL"]])
```
2. Drop the empty string to
```python
data = tf.RaggedTensor([["Hello", "world"], [], ["Welcome", "to", "ElasticDL"]])
```
3. Transform strings in the ragged tensor to integer ids by `Lookup` or `Hashing`
```python
data = tf.RaggedTensor([[0, 1], [], [2, 3, 4]])
```
4. Convert it to `Tensor` with padding by `tf.RaggedTensor.to_tensor(shape=(None, 3))`
```python
data = data.to_tensor(shape=(None,3))
tf.Tensor([[0, 1, 0], [0, 0 ,0], [2, 3, 4]])
```

### PR DESCRIPTION
**Pull request by workingloong**
Fix #1861

### CODE REVIEW

File: elasticdl_preprocessing/layers/to_ragged.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        inp = tf.constant([["1,2,3"], ["4,5"], [""]])
+        out = layer(inp)
+        [["1", "2", "3"], ["4", "5"], []]
+    ```
+
+    Arguments:
+        sep: Valid if the input's dtype is string.
+        ignore_value: Entries in inputs equal to this value will be
+            absent from the output `RaggedTensor`. If `None`, default value of
+            input's dtype will be used ('' for `str`, -1 for `int`).
```
Comment (terrytangyuan) [inline elasticdl_preprocessing/layers/to_ragged.py:33]: I had similar comment on this in the other PR

File: elasticdl_preprocessing/layers/to_ragged.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    Output shape: An `RaggedTensor` with the same shape as inputs
+    """
+
+    def __init__(self, sep=_COMMA_SEP, ignore_value=None):
+        super(ToRagged, self).__init__()
+        self.sep = sep
+        self.ignore_value = ignore_value
+
+    def call(self, inputs):
```
Comment (brightcoder01) [inline elasticdl_preprocessing/layers/to_ragged.py:46]: What will happen if inputs is a SparseTensor?

File: elasticdl_preprocessing/layers/to_ragged.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        if isinstance(inputs, tf.Tensor):
+            inputs = tf.RaggedTensor.from_tensor(inputs)
+
+        if self.ignore_value is None:
```
Comment (brightcoder01) [inline elasticdl_preprocessing/layers/to_ragged.py]: Same comments in the ToSparse layer

File: elasticdl_preprocessing/layers/to_ragged.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import tensorflow as tf
+
+_COMMA_SEP = ","
+
+
+class ToRagged(tf.keras.layers.Layer):
+    """Converts a `Tensor` or `RaggedTensor` to a `RaggedTensor`,
+    dropping ignore_value cells. If the input's dtype is string, split
+    the string elements to convert the input to `RaggedTensor` firstly.
```
Comment (brightcoder01) [inline elasticdl_preprocessing/layers/to_ragged.py:9]: What's the TF version requirement for this layer?

File: elasticdl_preprocessing/layers/to_ragged.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        [[1], [], [4]]
+    ```
+
+    Example (String):
+    ```python
+        ayer = ToRagged()
+        inp = tf.constant([["1,2,3"], ["4,5"], [""]])
+        out = layer(inp)
+    ```
+    The expected output is `[["1", "2", "3"], ["4", "5"], []]`
```
Comment (brightcoder01) [inline elasticdl_preprocessing/layers/to_ragged.py:27]: For string input, the rank of output will increase 1
For int input, the rank of output won't change.
It will be confusing to the users.

File: elasticdl_preprocessing/layers/to_ragged.py (hunk 2 above)
Comment (workingloong) [reply elasticdl_preprocessing/layers/to_ragged.py:46]: It may raise an unknown error. We need to check the input type and raise an error is the type is not `Tensor` or `RaggedTensor`.

File: elasticdl_preprocessing/layers/to_ragged.py (hunk 5 above)
Comment (workingloong) [reply elasticdl_preprocessing/layers/to_ragged.py:27]: So, is the expected output for string input `[[["1", "2", "3"]], [["4", "5"]], [[]]]`?

File: elasticdl_preprocessing/layers/to_ragged.py (hunk 4 above)
Comment (workingloong) [reply elasticdl_preprocessing/layers/to_ragged.py:9]: Above TF 2.0.0 and I have added a note.

File: elasticdl_preprocessing/layers/to_ragged.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    """
+
+    def __init__(self, sep=_COMMA_SEP, ignore_value=None):
+        super(ToRagged, self).__init__()
+        self.sep = sep
+        self.ignore_value = ignore_value
+
+    def call(self, inputs):
+        if not isinstance(inputs, tf.Tensor) and not isinstance(
+            inputs, tf.RaggedTensor
```
Comment (chunyang-wen) [inline elasticdl_preprocessing/layers/to_ragged.py]:
```python
if not isinstance(inputs, (tf.Tensor, tf.RaggedTensor):
    pass
```

File: elasticdl_preprocessing/layers/to_ragged.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        if isinstance(inputs, tf.Tensor):
+            inputs = tf.RaggedTensor.from_tensor(inputs)
+
+        ignore_value = self.ignore_value
+        if ignore_value is None:
+            if inputs.dtype == tf.string:
+                ignore_value = ""
+                inputs = tf.strings.split(inputs, sep=self.sep).values
+            elif inputs.dtype.is_integer:
+                ignore_value = -1
```
Comment (chunyang-wen) [inline elasticdl_preprocessing/layers/to_ragged.py]: what about other `dtype`? It seems that it is not complete. And you can add a process function for `ignore_value` instead of testing and setting it everywhere. All you have to do is to gather enough information and return a valid `ignore_value`.

File: elasticdl_preprocessing/tests/to_ragged_test.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import tensorflow as tf
+
+from elasticdl_preprocessing.layers.to_ragged import ToRagged
+from elasticdl_preprocessing.tests.test_utils import ragged_tensor_equal
+
+
+class ToRaggedTest(unittest.TestCase):
+    def test_dense_to_ragged(self):
+        layer = ToRagged()
+        inp = tf.constant([[1], [-1], [4]], tf.int64)
```
Comment (chunyang-wen) [inline elasticdl_preprocessing/tests/to_ragged_test.py]: Use abbreviations when they are commonly used, otherwise, you are making it hard to read the code. inp for input?

File: elasticdl_preprocessing/tests/to_ragged_test.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        out = layer(inp)
+        expected_out = tf.ragged.constant([[1], [], [4]], tf.int64)
+        self.assertTrue(ragged_tensor_equal(out, expected_out))
```
Comment (chunyang-wen) [inline elasticdl_preprocessing/tests/to_ragged_test.py:15]: I find that `RaggedTensor` has a member function `to_list`. You can convert the tensor to a numpy array and use a normal `assertListEqual` to compare them. I think this function will give more information when anything is wrong. BTW, I have not check the code for `ragged_tensor_equal`.

File: elasticdl_preprocessing/layers/to_ragged.py (hunk 7 above)
Comment (workingloong) [reply elasticdl_preprocessing/layers/to_ragged.py]: Yeah, Done!

File: elasticdl_preprocessing/tests/to_ragged_test.py (hunk 9 above)
Comment (workingloong) [reply elasticdl_preprocessing/tests/to_ragged_test.py:15]: `ragged_tensor_equal` checks the dtypes and contents of `RaggedTensor`. It is more simple to check the contents by `to_list` than the implementation in `ragged_tensor_equal`. I will use `to_list` to compare. Thank you for the suggestion!

**Review by brightcoder01 — APPROVED** [review]
